pkg/graph/parser/cel.go

```diff
@@ -49,19 +49,23 @@ func extractExpressions(str string) ([]string, error) {
 		// consider the outermost expression
 		bracketCount := 1
 		endIdx := startIdx + len(exprStart)
+
 		for endIdx < len(str) {
 			if str[endIdx] == '{' {
 				bracketCount++
 			} else if str[endIdx] == '}' {
 				bracketCount--
-				// If we have reached the end of the expression, break
 				if bracketCount == 0 {
 					break
 				}
 			} else if endIdx+1 < len(str) && str[endIdx:endIdx+2] == "${" {
-				// We do not allow nested expressions. I'm not sure if this is a
-				// good idea, but its sounds like a reasonable restriction.
-				return nil, ErrNestedExpression
+				// Allow nested expressions, but only if they are escaped with quotes, CEL will
+				// treat them as a string literal
+				if str[endIdx-1] != '"' {
+					// We do not allow nested expressions. I'm not sure if this is a
+					// good idea, but its sounds like a reasonable restriction.
+					return nil, ErrNestedExpression
+				}
```

Comment:
Thanks @dirien! I like the approach of allowing quoted nested expressions. one thing to consider though - the current implementation might have trouble with braces inside quoted strings. since we're counting all `{` and `}` characters regardless of whether they're in quotes. e.g an expression like `${"text with } inside"}` could end prematurely, and we'll only extract "`text with }`" as a CEL expression.

I'm thinking that  we could track whether we're inside a string literal as we parse? That way we'd only count braces when they're actually structural elements of the CEL expression

Reply:
@a-hilaly made the changes. Have a look if you like it.